js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataThread.tsx

```diff
@@ -0,0 +1,255 @@
+import {ApolloClient, gql} from '@apollo/client';
+
+import {LiveDataForNode, tokenForAssetKey, buildLiveDataForNode} from '../asset-graph/Utils';
+import {AssetKeyInput} from '../graphql/types';
+
+import type {AssetLiveDataThreadManager} from './AssetLiveDataThreadManager';
+import {
+  AssetGraphLiveQuery,
+  AssetGraphLiveQueryVariables,
+  AssetNodeLiveFragment,
+} from './types/AssetLiveDataThread.types';
+import {BATCHING_INTERVAL} from './util';
+
+export type AssetLiveDataThreadID = 'default' | 'sidebar' | 'asset-graph';
+
+export class AssetLiveDataThread {
+  private isFetching: boolean = false;
+  private listenersCount: {[key: string]: number};
+  private client: ApolloClient<any>;
+  private isLooping: boolean = false;
+  private interval?: ReturnType<typeof setTimeout>;
+  private manager: AssetLiveDataThreadManager;
+  public pollRate: number = 30000;
+
+  protected static _threads: {[key: string]: AssetLiveDataThread} = {};
+
+  constructor(client: ApolloClient<any>, manager: AssetLiveDataThreadManager) {
+    this.client = client;
+    this.listenersCount = {};
+    this.manager = manager;
+  }
+
+  public setPollRate(pollRate: number) {
+    this.pollRate = pollRate;
+  }
+
+  public subscribe(key: string) {
+    this.listenersCount[key] = this.listenersCount[key] || 0;
+    this.listenersCount[key] += 1;
+    this.startFetchLoop();
+  }
+
+  public unsubscribe(key: string) {
+    this.listenersCount[key] -= 1;
+    if (this.listenersCount[key] === 0) {
+      delete this.listenersCount[key];
+    }
+    if (this.getObservedKeys().length === 0) {
+      this.stopFetchLoop();
+    }
+  }
+
+  public getObservedKeys() {
+    return Object.keys(this.listenersCount);
+  }
+
+  public startFetchLoop() {
+    if (this.isLooping) {
+      return;
+    }
+    this.isLooping = true;
+    const fetch = () => {
+      const nextAssets = this.manager.determineAssetsToFetch(this.getObservedKeys());
+      if (nextAssets.length) {
+        this._batchedQueryAssets(nextAssets);
+      }
+    };
+    setTimeout(fetch, BATCHING_INTERVAL);
+    this.interval = setInterval(fetch, 5000);
+  }
+
+  public stopFetchLoop() {
+    if (!this.isLooping) {
+      return;
+    }
+    this.isLooping = false;
+    clearInterval(this.interval);
+    this.interval = undefined;
+  }
+
+  private async queryAssetKeys(assetKeys: AssetKeyInput[]) {
+    const {data} = await this.client.query<AssetGraphLiveQuery, AssetGraphLiveQueryVariables>({
+      query: ASSETS_GRAPH_LIVE_QUERY,
+      fetchPolicy: 'network-only',
+      variables: {
+        assetKeys,
+      },
+    });
+    const nodesByKey: Record<string, AssetNodeLiveFragment> = {};
+    const liveDataByKey: Record<string, LiveDataForNode> = {};
+    data.assetNodes.forEach((assetNode) => {
```

Comment:
Purely stylistic, you want you could re-write this as `const nodesByKey = Object.fromEntries(data.assetNodes.map(node => [tokenForAssetKey(node.assetKey), node]))`. I've been really liking that syntax because the types on the returned object are auto-inferred.

Reply:
yeah good point, this occurred to me while writing this too, let me change it